src/main/java/taiga/api/APIWrapperBase.java

```diff
@@ -146,28 +153,54 @@ protected <T> CompletableFuture<APIResponse<T>> queryAsync(
                 request.header("Authorization", "Bearer " + authToken);
             }
 
-            return HTTPClientSingleton.getInstance()
-                    .sendAsync(request.GET().build(), HttpResponse.BodyHandlers.ofString())
+            Semaphore semaphore = semaphores.computeIfAbsent(getRequestKey(request.build()), k -> new Semaphore(MAX_CONCURRENT_REQUESTS_NUMBER));
+
+            // Acquire a permit from the semaphore
+            semaphore.acquireUninterruptibly();
+    
+            return CompletableFuture.supplyAsync(() -> {
+                        try {
+                            return HTTPClientSingleton.getInstance()
+                                    .sendAsync(request.GET().build(), HttpResponse.BodyHandlers.ofString())
+                                    .thenApply(response -> {
+                                        AtomicReference<APIResponse<T>> apiResponse =
+                                                new AtomicReference<>(createResponse(response, responseType));
+    
+                                        if (retry && apiResponse.get().getStatus() == 401) {
+                                            refreshAuthToken(query, responseType, apiResponse, retry, enable_pagination);
+                                        } else if (apiResponse.get().getStatus() == 429) {
```

Comment:
I have doubts on whether this is synonymous with the concurrent streams issue. My reasoning is that the status code is set from the taiga server, so the taiga API could suddenly decide any client invoking greater than 10 requests should receive a 429, when we've definitely handled more concurrent requests.